ERCS/erc-7746.md

```diff
@@ -0,0 +1,97 @@
+
+## Specification
+
+A contract implementing the `ILayer` interface MUST provide two functions:
+
+```solidity
+// SPDX-License-Identifier: CC0-1.0
+pragma solidity 0.8.20;
+
+interface ILayer {
+    /// @notice Validates a function call before execution.
+    /// @param layerConfig Layer-specific configuration data.
+    /// @param selector The function selector being called.
+    /// @param sender The address initiating the call.
+    /// @param value The amount of ETH sent with the call (if any).
+    /// @param data The calldata for the function call.
+    /// @return beforeCallResult Arbitrary data to be passed to `afterCallValidation`.
+    /// @dev MUST revert if validation fails.
+    function beforeCallValidation(
+        bytes memory layerConfig,
+        bytes4 selector,
+        address sender,
+        uint256 value,
+        bytes memory data
+    ) external returns (bytes memory);
+
+    /// @notice Validates a function call after execution.
+    /// @param layerConfig Layer-specific configuration data.
+    /// @param selector The function selector being called.
+    /// @param sender The address initiating the call.
+    /// @param value The amount of ETH sent with the call (if any).
+    /// @param data The calldata for the function call.
+    /// @param beforeCallResult The data returned by `beforeCallValidation`.
+    /// @dev MUST revert if validation fails.
+    function afterCallValidation(
+        bytes memory layerConfig,
+        bytes4 selector,
+        address sender,
+        uint256 value,
+        bytes memory data,
+        bytes memory beforeCallResult
+    ) external;
+}
+
+```
+
+A protected contract MAY integrate security layers by calling the beforeCallValidation function before executing its logic and the afterCallValidation function afterwards. Multiple layers can be registered and executed in a defined order. The protected contract MUST revert if any layer reverts.
+
+## Rationale
+
+**Flexibility**: The `layerConfig` parameter allows for layer-specific customization, enabling a single layer implementation to serve multiple contracts with varying requirements.
+
+**non-static calls**: Layers can maintain their own state, allowing for more complex validation logic (e.g., rate limiting, usage tracking).
+
+**Strict Validation**: Reverts on validation failure ensure a fail-safe mechanism, preventing execution of potentially harmful transactions.
+
+**Gas Costs**: Layers naturally will have gas costs associated with their execution. However, the benefits of enhanced security and modularity outweigh these costs, especially as blockchain technology continues to evolve and we expect gas costs to decrease over time.
+
+## Reference Implementation
+
+A reference implementation of the ILayer interface and a sample protected contract can be found in the repository:
+[../assets/erc-7746/](../assets/erc-7746/)
+
```

Comment:
I'd like to see a bit of a walkthrough of the reference implementation. For example, which files are important to the standard, and which are just there for testing.